### ISSUE TITLE
Authentication fails with "AADSTS50097: Device authentication is required."

### ISSUE THREAD

**Issue by hruan** [microsoft/artifacts-credprovider#204]
`dotnet restore --interactive` fails and debug log shows an exception from ADAL with code AADSTS50097.

[This issue](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/issues/1543) with ADAL suggest that this is/will be fixed in MSAL. As [MSAL obsoletes ADAL](https://docs.microsoft.com/en-us/azure/active-directory/develop/msal-migration#frequently-asked-questions-faq), is a migration to MSAL in the plans?

```
Restoring packages for .NETCoreApp,Version=v3.1...
Using C:\Users\...\.nuget\plugins\netcore\CredentialProvider.Microsoft\CredentialProvider.Microsoft.dll as a credential provider plugin.
    [CredentialProvider.091812]Running in plug-in mode
    [CredentialProvider.091812]Command-line v0.1.24+de2b62bcfccebababa9ea4099aba6c44d2f349a4: C:\Users\..\.nuget\plugins\netcore\CredentialProvider.Microsoft\CredentialProvider.Microsoft.dll -Plugin
    [CredentialProvider.091812]Handling 'Request' 'Initialize'. Time elapsed in ms: 8 - Payload: {"ClientVersion":"5.6.0","Culture":"en-SE","RequestTimeout":"00:00:05"}
    [CredentialProvider.091812]Sending response: 'Request' 'Initialize'. Time elapsed in ms: 10
    [CredentialProvider.091812]Time elapsed in milliseconds after sending response 'Request' 'Initialize': 12
    [CredentialProvider.091812]Handling 'Request' 'GetOperationClaims'. Time elapsed in ms: 1 - Payload: {}
    [CredentialProvider.091812]Sending response: 'Request' 'GetOperationClaims'. Time elapsed in ms: 10
    [CredentialProvider.091812]Time elapsed in milliseconds after sending response 'Request' 'GetOperationClaims': 34
    [CredentialProvider.091812]Handling 'Request' 'SetLogLevel'. Time elapsed in ms: 2 - Payload: {"LogLevel":"Debug"}
    [CredentialProvider]Sending response: 'Request' 'SetLogLevel'. Time elapsed in ms: 4
    [CredentialProvider]Time elapsed in milliseconds after sending response 'Request' 'SetLogLevel': 10
    [CredentialProvider]Handling 'Request' 'GetAuthenticationCredentials'. Time elapsed in ms: 10 - Payload: {"Uri":"https://pkgs.dev.azure.com/.../nuget/v3/index.json","IsRetry":false,"IsNonInteractive":false,"CanShowDialog":false}
    [CredentialProvider]Creating a progress reporter with interval: 00:00:02
    [CredentialProvider]Handling auth request, Uri: https://pkgs.dev.azure.com/.../nuget/v3/index.json, IsRetry: False, IsNonInteractive: False, CanShowDialog: False
    [CredentialProvider]URI: https://pkgs.dev.azure.com/.../nuget/v3/index.json
    [CredentialProvider]VstsBuildTaskServiceEndpointCredentialProvider - This credential provider must be run under the Team Build tasks for NuGet with external endpoint credentials. Appropriate environment variable needs to be set.
    [CredentialProvider]Skipping NuGetCredentialProvider.CredentialProviders.VstsBuildTaskServiceEndpoint.VstsBuildTaskServiceEndpointCredentialProvider, cannot provide credentials for https://pkgs.dev.azure.com/.../nuget/v3/index.json
    [CredentialProvider]VstsBuildTaskCredentialProvider - This credential provider must be run under the Team Build tasks for NuGet. Appropriate environment variables must be set.
    [CredentialProvider]Skipping NuGetCredentialProvider.CredentialProviders.VstsBuildTask.VstsBuildTaskCredentialProvider, cannot provide credentials for https://pkgs.dev.azure.com/.../nuget/v3/index.json
    [CredentialProvider]GET https://pkgs.dev.azure.com/.../nuget/v3/index.json
    [CredentialProvider]VstsCredentialProvider - Detected a hosted Azure DevOps Service.
    [CredentialProvider]Using NuGetCredentialProvider.CredentialProviders.Vsts.VstsCredentialProvider to try to get credentials for https://pkgs.dev.azure.com/.../nuget/v3/index.json.
    [CredentialProvider]IsRetry: False
    [CredentialProvider]Could not find cached SessionToken for https://pkgs.dev.azure.com/.../nuget/v3/index.json
    [CredentialProvider]GET https://pkgs.dev.azure.com/.../nuget/v3/index.json
    [CredentialProvider]Found AAD Authority from 401 headers: https://login.windows.net/...
    [CredentialProvider]VstsCredentialProvider - Using AAD authority: https://login.windows.net/...
    [CredentialProvider]VstsCredentialProvider - Attempting to acquire bearer token using provider 'ADAL Cache'

    [CredentialProvider]VstsCredentialProvider - Bearer token provider 'ADAL Cache' didn't acquire a token
    [CredentialProvider]VstsCredentialProvider - Attempting to acquire bearer token using provider 'ADAL Windows Integrated Authentication'
    [CredentialProvider]VstsCredentialProvider - Bearer token provider 'ADAL Windows Integrated Authentication' failed with exception:\nMicrosoft.IdentityModel.Clients.ActiveDirectory.AdalException: Integrated Windows Auth is not supported for managed users. See https://aka.ms/adal-iwa for details.
   at Microsoft.IdentityModel.Clients.ActiveDirectory.Internal.Flows.AcquireTokenIWAHandler.PreTokenRequestAsync()
   at Microsoft.IdentityModel.Clients.ActiveDirectory.Internal.Flows.AcquireTokenHandlerBase.RunAsync()
   at Microsoft.IdentityModel.Clients.ActiveDirectory.AuthenticationContext.AcquireTokenCommonAsync(String resource, String clientId, IntegratedWindowsAuthInput iwaInput)
   at Microsoft.IdentityModel.Clients.ActiveDirectory.AuthenticationContextIntegratedAuthExtensions.AcquireTokenAsync(AuthenticationContext ctx, String resource, String clientId, UserCredential userCredential)
   at NuGetCredentialProvider.CredentialProviders.Vsts.AdalTokenProvider.AcquireTokenWithWindowsIntegratedAuth(CancellationToken cancellationToken) in E:\A\_work\919\s\CredentialProvider.Microsoft\CredentialProviders\Vsts\AdalTokenProvider.cs:line 122
   at NuGetCredentialProvider.CredentialProviders.Vsts.WindowsIntegratedAuthBearerTokenProvider.GetTokenAsync(Uri uri, CancellationToken cancellationToken) in E:\A\_work\919\s\CredentialProvider.Microsoft\CredentialProviders\Vsts\BearerTokenProviders.cs:line 58
   at NuGetCredentialProvider.CredentialProviders.Vsts.VstsCredentialProvider.HandleRequestAsync(GetAuthenticationCredentialsRequest request, CancellationToken cancellationToken) in E:\A\_work\919\s\CredentialProvider.Microsoft\CredentialProviders\Vsts\VstsCredentialProvider.cs:line 117
       ErrorCode: integrated_windows_auth_not_supported_managed_user
    [CredentialProvider]VstsCredentialProvider - Not running bearer token provider 'ADAL UI'
    [CredentialProvider]VstsCredentialProvider - Attempting to acquire bearer token using provider 'ADAL Device Code'
    [CredentialProvider]DeviceFlow: https://pkgs.dev.azure.com/.../nuget/v3/index.json
    [CredentialProvider]ATTENTION: User interaction required.

    **********************************************************************

    To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code ... to authenticate.

    **********************************************************************

    [CredentialProvider]VstsCredentialProvider - Bearer token provider 'ADAL Device Code' failed with exception:\nMicrosoft.IdentityModel.Clients.ActiveDirectory.AdalServiceException: AADSTS50097: Device authentication is required.

Trace ID: ...
Correlation ID: ...
Timestamp: 2020-07-31 07:18:59Z ---> Microsoft.IdentityModel.Clients.ActiveDirectory.AdalServiceException: Response status code does not indicate success: 401 (Unauthorized).
   at Microsoft.Identity.Core.OAuth2.OAuthClient.GetResponseAsync[T](Boolean respondToDeviceAuthChallenge)
   --- End of inner exception stack trace ---
   at Microsoft.Identity.Core.OAuth2.OAuthClient.GetResponseAsync[T](Boolean respondToDeviceAuthChallenge)
   at Microsoft.Identity.Core.OAuth2.OAuthClient.GetResponseAsync[T]()
   at Microsoft.IdentityModel.Clients.ActiveDirectory.Internal.Flows.AcquireTokenHandlerBase.SendHttpMessageAsync(IRequestParameters requestParameters)
   at Microsoft.IdentityModel.Clients.ActiveDirectory.Internal.Flows.AcquireTokenHandlerBase.SendTokenRequestAsync()
   at Microsoft.IdentityModel.Clients.ActiveDirectory.Internal.Flows.AcquireTokenByDeviceCodeHandler.SendTokenRequestAsync()
   at Microsoft.IdentityModel.Clients.ActiveDirectory.Internal.Flows.AcquireTokenHandlerBase.CheckAndAcquireTokenUsingBrokerAsync()
   at Microsoft.IdentityModel.Clients.ActiveDirectory.Internal.Flows.AcquireTokenHandlerBase.RunAsync()
   at Microsoft.IdentityModel.Clients.ActiveDirectory.AuthenticationContext.AcquireTokenByDeviceCodeAsync(DeviceCodeResult deviceCodeResult, CancellationToken cancellationToken)
   at NuGetCredentialProvider.CredentialProviders.Vsts.AdalTokenProvider.AcquireTokenWithDeviceFlowAsync(Func`2 deviceCodeHandler, CancellationToken cancellationToken, ILogger logger) in E:\A\_work\919\s\CredentialProvider.Microsoft\CredentialProviders\Vsts\AdalTokenProvider.cs:line 52
   at NuGetCredentialProvider.CredentialProviders.Vsts.DeviceCodeFlowBearerTokenProvider.GetTokenAsync(Uri uri, CancellationToken cancellationToken) in E:\A\_work\919\s\CredentialProvider.Microsoft\CredentialProviders\Vsts\BearerTokenProviders.cs:line 122
   at NuGetCredentialProvider.CredentialProviders.Vsts.VstsCredentialProvider.HandleRequestAsync(GetAuthenticationCredentialsRequest request, CancellationToken cancellationToken) in E:\A\_work\919\s\CredentialProvider.Microsoft\CredentialProviders\Vsts\VstsCredentialProvider.cs:line 117
       ErrorCode: invalid_grant
       StatusCode: 401
    [CredentialProvider]VstsCredentialProvider - Could not obtain credentials for https://pkgs.dev.azure.com/.../nuget/v3/index.json
    [CredentialProvider]Unable to acquire credentials.
    [CredentialProvider]Sending response: 'Request' 'GetAuthenticationCredentials'. Time elapsed in ms: 47486
    [CredentialProvider]Time elapsed in milliseconds after sending response 'Request' 'GetAuthenticationCredentials': 47491
    1>C:\Program Files\dotnet\sdk\3.1.301\NuGet.targets(128,5): error : Unable to load the service index for source https://pkgs.dev.azure.com/.../nuget/v3/index.json. [C:\_Repos\...sln]
C:\Program Files\dotnet\sdk\3.1.301\NuGet.targets(128,5): error :   Response status code does not indicate success: 401 (Unauthorized). [C:\_Repos\....sln]
    NuGet.Protocol.Core.Types.FatalProtocolException: Unable to load the service index for source https://pkgs.dev.azure.com/.../nuget/v3/index.json.
     ---> System.Net.Http.HttpRequestException: Response status code does not indicate success: 401 (Unauthorized).

       at System.Net.Http.HttpResponseMessage.EnsureSuccessStatusCode()
       at NuGet.Protocol.HttpSource.<>c__DisplayClass14_0`1.<<GetAsync>b__0>d.MoveNext()
    --- End of stack trace from previous location where exception was thrown ---
       at NuGet.Common.ConcurrencyUtilities.ExecuteWithFileLockedAsync[T](String filePath, Func`2 action, CancellationToken token)
       at NuGet.Common.ConcurrencyUtilities.ExecuteWithFileLockedAsync[T](String filePath, Func`2 action, CancellationToken token)
       at NuGet.Protocol.HttpSource.GetAsync[T](HttpSourceCachedRequest request, Func`2 processAsync, ILogger log, CancellationToken token)
       at NuGet.Protocol.ServiceIndexResourceV3Provider.GetServiceIndexResourceV3(SourceRepository source, DateTime utcNow, ILogger log, CancellationToken token)
       --- End of inner exception stack trace ---
       at NuGet.Protocol.ServiceIndexResourceV3Provider.GetServiceIndexResourceV3(SourceRepository source, DateTime utcNow, ILogger log, CancellationToken token)
       at NuGet.Protocol.ServiceIndexResourceV3Provider.TryCreate(SourceRepository source, CancellationToken token)

       at NuGet.Protocol.Core.Types.SourceRepository.GetResourceAsync[T](CancellationToken token)
       at NuGet.Protocol.Core.Types.SourceRepository.GetResourceAsync[T]()
       at NuGet.Protocol.HttpFileSystemBasedFindPackageByIdResourceProvider.TryCreate(SourceRepository sourceRepository, CancellationToken token)
       at NuGet.Protocol.Core.Types.SourceRepository.GetResourceAsync[T](CancellationToken token)
       at NuGet.Protocol.Core.Types.SourceRepository.GetResourceAsync[T]()
       at NuGet.Commands.SourceRepositoryDependencyProvider.EnsureResource()
       at NuGet.Commands.SourceRepositoryDependencyProvider.FindLibraryCoreAsync(LibraryRange libraryRange, SourceCacheContext cacheContext, ILogger logger, CancellationToken cancellationToken)
       at NuGet.Commands.SourceRepositoryDependencyProvider.<>c__DisplayClass19_0.<<FindLibraryAsync>b__0>d.MoveNext()
    --- End of stack trace from previous location where exception was thrown ---
       at NuGet.Commands.SourceRepositoryDependencyProvider.FindLibraryAsync(LibraryRange libraryRange, NuGetFramework targetFramework, SourceCacheContext cacheContext, ILogger logger, CancellationToken cancellationToken)
       at NuGet.DependencyResolver.ResolverUtility.<>c__DisplayClass9_1.<<FindLibraryFromSourcesAsync>b__0>d.MoveNext()
    --- End of stack trace from previous location where exception was thrown ---
       at NuGet.DependencyResolver.ResolverUtility.FindLibraryFromSourcesAsync(LibraryRange libraryRange, IEnumerable`1 providers, Func`2 action)
       at NuGet.DependencyResolver.ResolverUtility.FindLibraryByVersionAsync(LibraryRange libraryRange, NuGetFramework framework, IEnumerable`1 providers, SourceCacheContext cacheContext, ILogger logger, CancellationToken token)
       at NuGet.DependencyResolver.ResolverUtility.FindPackageLibraryMatchAsync(LibraryRange libraryRange, NuGetFramework framework, IEnumerable`1 remoteProviders, IEnumerable`1 localProviders, SourceCacheContext cacheContext, ILogger logger, CancellationToken cancellationToken)
       at NuGet.DependencyResolver.ResolverUtility.FindLibraryMatchAsync(LibraryRange libraryRange, NuGetFramework framework, String runtimeIdentifier, IEnumerable`1 remoteProviders, IEnumerable`1 localProviders, IEnumerable`1 projectProviders, IDictionary`2 lockFileLibraries, SourceCacheContext cacheContext, ILogger logger, CancellationToken cancellationToken)
       at NuGet.DependencyResolver.ResolverUtility.FindLibraryEntryAsync(LibraryRange libraryRange, NuGetFramework framework, String runtimeIdentifier, RemoteWalkContext context, CancellationToken cancellationToken)
       at NuGet.DependencyResolver.RemoteDependencyWalker.CreateGraphNode(LibraryRange libraryRange, NuGetFramework framework, String runtimeName, RuntimeGraph runtimeGraph, Func`2 predicate, GraphEdge`1 outerEdge, TransitiveCentralPackageVersions transitiveCentralPackageVersions)
       at NuGet.DependencyResolver.RemoteDependencyWalker.CreateGraphNode(LibraryRange libraryRange, NuGetFramework framework, String runtimeName, RuntimeGraph runtimeGraph, Func`2 predicate, GraphEdge`1 outerEdge, TransitiveCentralPackageVersions transitiveCentralPackageVersions)
       at NuGet.DependencyResolver.RemoteDependencyWalker.CreateGraphNode(LibraryRange libraryRange, NuGetFramework framework, String runtimeName, RuntimeGraph runtimeGraph, Func`2 predicate, GraphEdge`1 outerEdge, TransitiveCentralPackageVersions transitiveCentralPackageVersions)
       at NuGet.DependencyResolver.RemoteDependencyWalker.WalkAsync(LibraryRange library, NuGetFramework framework, String runtimeIdentifier, RuntimeGraph runtimeGraph, Boolean recursive)
       at NuGet.Commands.ProjectRestoreCommand.WalkDependenciesAsync(LibraryRange projectRange, NuGetFramework framework, String runtimeIdentifier, RuntimeGraph runtimeGraph, RemoteDependencyWalker walker, RemoteWalkContext context, CancellationToken token)
       at NuGet.Commands.ProjectRestoreCommand.TryRestoreAsync(LibraryRange projectRange, IEnumerable`1 frameworkRuntimePairs, NuGetv3LocalRepository userPackageFolder, IReadOnlyList`1 fallbackPackageFolders, RemoteDependencyWalker remoteWalker, RemoteWalkContext context, Boolean forceRuntimeGraphCreation, CancellationToken token, TelemetryActivity telemetryActivity)
       at NuGet.Commands.RestoreCommand.ExecuteRestoreAsync(NuGetv3LocalRepository userPackageFolder, IReadOnlyList`1 fallbackPackageFolders, RemoteWalkContext context, CancellationToken token, TelemetryActivity telemetryActivity)
       at NuGet.Commands.RestoreCommand.ExecuteAsync(CancellationToken token)
       at NuGet.Commands.RestoreRunner.ExecuteAsync(RestoreSummaryRequest summaryRequest, CancellationToken token)
       at NuGet.Commands.RestoreRunner.ExecuteAndCommitAsync(RestoreSummaryRequest summaryRequest, CancellationToken token)
  Done executing task "RestoreTask" -- FAILED.
1>Done building target "Restore" in project "....sln" -- FAILED.
1>Done Building Project "C:\_Repos\....sln" (Restore target(s)) -- FAILED.

Build FAILED.
```

**Comment by nayanshah** [comment]
Adding MSAL support seems to be already in progress - #185 

But the [ADAL issue](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/issues/1543) linked above only suggests that MSAL will improve the message for why Device Code auth failed. Actual error might be that username/password auth or MFA is required but that's not shown because credprovider is running in non-interactive mode. 

Try running the credprovider with `-C -R -Uri <feed uri>` to enable showing UI.

**Comment by hruan** [comment]
No such luck, I'm afraid.

```
[Verbose] [CredentialProvider]Running in stand-alone mode
[Verbose] [CredentialProvider]Command-line v0.1.24+de2b62bcfccebababa9ea4099aba6c44d2f349a4: C:\Users\...\.nuget\plugins\netcore\CredentialProvider.Microsoft\CredentialProvider.Microsoft.dll -R -C -Uri https://pkgs.dev.azure.com/.../nuget/v3/index.json -V Debug
[Verbose] [CredentialProvider]Handling auth request, Uri: https://pkgs.dev.azure.com/.../nuget/v3/index.json, IsRetry: False, IsNonInteractive: False, CanShowDialog: True
[Verbose] [CredentialProvider]URI: https://pkgs.dev.azure.com/.../nuget/v3/index.json
[Verbose] [CredentialProvider]VstsBuildTaskServiceEndpointCredentialProvider - This credential provider must be run under the Team Build tasks for NuGet with external endpoint credentials. Appropriate environment variable needs to be set.
[Verbose] [CredentialProvider]Skipping NuGetCredentialProvider.CredentialProviders.VstsBuildTaskServiceEndpoint.VstsBuildTaskServiceEndpointCredentialProvider, cannot provide credentials for https://pkgs.dev.azure.com/.../nuget/v3/index.json
[Verbose] [CredentialProvider]VstsBuildTaskCredentialProvider - This credential provider must be run under the Team Build tasks for NuGet. Appropriate environment variables must be set.
[Verbose] [CredentialProvider]Skipping NuGetCredentialProvider.CredentialProviders.VstsBuildTask.VstsBuildTaskCredentialProvider, cannot provide credentials for https://pkgs.dev.azure.com/.../nuget/v3/index.json
[Verbose] [CredentialProvider]GET https://pkgs.dev.azure.com/.../nuget/v3/index.json
[Verbose] [CredentialProvider]VstsCredentialProvider - Detected a hosted Azure DevOps Service.
[Verbose] [CredentialProvider]Using NuGetCredentialProvider.CredentialProviders.Vsts.VstsCredentialProvider to try to get credentials for https://pkgs.dev.azure.com/.../nuget/v3/index.json.
[Verbose] [CredentialProvider]IsRetry: False
[Verbose] [CredentialProvider]Could not find cached SessionToken for https://pkgs.dev.azure.com/.../nuget/v3/index.json
[Verbose] [CredentialProvider]GET https://pkgs.dev.azure.com/.../nuget/v3/index.json
[Verbose] [CredentialProvider]Found AAD Authority from 401 headers: https://login.windows.net/...
[Verbose] [CredentialProvider]VstsCredentialProvider - Using AAD authority: https://login.windows.net/...
[Verbose] [CredentialProvider]VstsCredentialProvider - Attempting to acquire bearer token using provider 'ADAL Cache'
[Verbose] [CredentialProvider]VstsCredentialProvider - Bearer token provider 'ADAL Cache' didn't acquire a token
[Verbose] [CredentialProvider]VstsCredentialProvider - Attempting to acquire bearer token using provider 'ADAL Windows Integrated Authentication'
[Verbose] [CredentialProvider]VstsCredentialProvider - Bearer token provider 'ADAL Windows Integrated Authentication' failed with exception:\nMicrosoft.IdentityModel.Clients.ActiveDirectory.AdalException: Integrated Windows Auth is not supported for managed users. See https://aka.ms/adal-iwa for details.
   at Microsoft.IdentityModel.Clients.ActiveDirectory.Internal.Flows.AcquireTokenIWAHandler.PreTokenRequestAsync()
   at Microsoft.IdentityModel.Clients.ActiveDirectory.Internal.Flows.AcquireTokenHandlerBase.RunAsync()
   at Microsoft.IdentityModel.Clients.ActiveDirectory.AuthenticationContext.AcquireTokenCommonAsync(String resource, String clientId, IntegratedWindowsAuthInput iwaInput)
   at Microsoft.IdentityModel.Clients.ActiveDirectory.AuthenticationContextIntegratedAuthExtensions.AcquireTokenAsync(AuthenticationContext ctx, String resource, String clientId, UserCredential userCredential)
   at NuGetCredentialProvider.CredentialProviders.Vsts.AdalTokenProvider.AcquireTokenWithWindowsIntegratedAuth(CancellationToken cancellationToken) in E:\A\_work\919\s\CredentialProvider.Microsoft\CredentialProviders\Vsts\AdalTokenProvider.cs:line 122
   at NuGetCredentialProvider.CredentialProviders.Vsts.WindowsIntegratedAuthBearerTokenProvider.GetTokenAsync(Uri uri, CancellationToken cancellationToken) in E:\A\_work\919\s\CredentialProvider.Microsoft\CredentialProviders\Vsts\BearerTokenProviders.cs:line 58
   at NuGetCredentialProvider.CredentialProviders.Vsts.VstsCredentialProvider.HandleRequestAsync(GetAuthenticationCredentialsRequest request, CancellationToken cancellationToken) in E:\A\_work\919\s\CredentialProvider.Microsoft\CredentialProviders\Vsts\VstsCredentialProvider.cs:line 117
        ErrorCode: integrated_windows_auth_not_supported_managed_user
[Verbose] [CredentialProvider]VstsCredentialProvider - Not running bearer token provider 'ADAL UI'
[Verbose] [CredentialProvider]VstsCredentialProvider - Attempting to acquire bearer token using provider 'ADAL Device Code'
[Minimal] [CredentialProvider]DeviceFlow: https://pkgs.dev.azure.com/.../nuget/v3/index.json
[Minimal] [CredentialProvider]ATTENTION: User interaction required.

    **********************************************************************

    To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code ... to authenticate.

    **********************************************************************

[Verbose] [CredentialProvider]VstsCredentialProvider - Bearer token provider 'ADAL Device Code' failed with exception:\nMicrosoft.IdentityModel.Clients.ActiveDirectory.AdalServiceException: AADSTS50097: Device authentication is required.
Trace ID: ...
Correlation ID: ...
Timestamp: 2020-08-04 06:12:06Z ---> Microsoft.IdentityModel.Clients.ActiveDirectory.AdalServiceException: Response status code does not indicate success: 401 (Unauthorized).
   at Microsoft.Identity.Core.OAuth2.OAuthClient.GetResponseAsync[T](Boolean respondToDeviceAuthChallenge)
   --- End of inner exception stack trace ---
   at Microsoft.Identity.Core.OAuth2.OAuthClient.GetResponseAsync[T](Boolean respondToDeviceAuthChallenge)
   at Microsoft.Identity.Core.OAuth2.OAuthClient.GetResponseAsync[T]()
   at Microsoft.IdentityModel.Clients.ActiveDirectory.Internal.Flows.AcquireTokenHandlerBase.SendHttpMessageAsync(IRequestParameters requestParameters)
   at Microsoft.IdentityModel.Clients.ActiveDirectory.Internal.Flows.AcquireTokenHandlerBase.SendTokenRequestAsync()
   at Microsoft.IdentityModel.Clients.ActiveDirectory.Internal.Flows.AcquireTokenByDeviceCodeHandler.SendTokenRequestAsync()
   at Microsoft.IdentityModel.Clients.ActiveDirectory.Internal.Flows.AcquireTokenHandlerBase.CheckAndAcquireTokenUsingBrokerAsync()
   at Microsoft.IdentityModel.Clients.ActiveDirectory.Internal.Flows.AcquireTokenHandlerBase.RunAsync()
   at Microsoft.IdentityModel.Clients.ActiveDirectory.AuthenticationContext.AcquireTokenByDeviceCodeAsync(DeviceCodeResult deviceCodeResult, CancellationToken cancellationToken)
   at NuGetCredentialProvider.CredentialProviders.Vsts.AdalTokenProvider.AcquireTokenWithDeviceFlowAsync(Func`2 deviceCodeHandler, CancellationToken cancellationToken, ILogger logger) in E:\A\_work\919\s\CredentialProvider.Microsoft\CredentialProviders\Vsts\AdalTokenProvider.cs:line 52
   at NuGetCredentialProvider.CredentialProviders.Vsts.DeviceCodeFlowBearerTokenProvider.GetTokenAsync(Uri uri, CancellationToken cancellationToken) in E:\A\_work\919\s\CredentialProvider.Microsoft\CredentialProviders\Vsts\BearerTokenProviders.cs:line 122
   at NuGetCredentialProvider.CredentialProviders.Vsts.VstsCredentialProvider.HandleRequestAsync(GetAuthenticationCredentialsRequest request, CancellationToken cancellationToken) in E:\A\_work\919\s\CredentialProvider.Microsoft\CredentialProviders\Vsts\VstsCredentialProvider.cs:line 117
        ErrorCode: invalid_grant
        StatusCode: 401
[Verbose] [CredentialProvider]VstsCredentialProvider - Could not obtain credentials for https://pkgs.dev.azure.com/.../nuget/v3/index.json
[Verbose] [CredentialProvider]Unable to acquire credentials.
```

**Comment by nayanshah** [comment]
The UI isn't enabled in .NET Core version of the credential provider. Could you try running the .NET Framework version instead? 

`
#if NETFRAMEWORK
            return !isNonInteractive && canShowDialog && RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
#else
            // no ADAL UI in netcore
            return false;
#endif
`

**Comment by hruan** [comment]
Using the netfx version of the credentials provider is a workaround for those running on Windows. However, the issue as I wrote in my first post the issue is really with `dotnet restore` and the netcoreapp version of the provider. I guess this still is an issue for those running Linux or macOS.

As a side note, I ran into this issue (at least one that had the same error code from Azure AD) with the Azure modules for Powershell. The discussions in [this issue](https://github.com/Azure/azure-powershell/issues/9427) suggested that it was an ADAL issue that was resolved in the preview version the account module that uses MSAL.

I tried the preview account module this morning and it worked as expected.

So, I think there is a good chance the issue with `dotnet restore` could be resolved if the three-month-old PR #185 was revived and merged. :)
